### DISCUSSION TITLE
How can I read the status of a pin via gcode?

### DISCUSSION THREAD

**Discussion by tcurdt** [grblHAL/core#519] [Q&A] [answered]
I've set the mask to include everything - but there is no information on input pin states

```
$10
$10=4095 (Status report options, mask)
ok
?
<Idle|MPos:0.000,0.000,0.000|Bf:128,1024|FS:0,0,0>
```

**Comment by terjeio** [comment, marked as answer]
> How can I read the status of a pin via gcode?

It is not possible unless it is an auxillary input. Auxillary inputs can be read via M66 and the result is stored in parameter 5399. To read and act upon parameter 5399 you will have to enable [expression support](https://github.com/grblHAL/core/wiki/Expressions-and-flow-control) in the controller.

A plugin could be written that allows accessing limit- and control inputs via custom M-code.

> I've set the mask to include everything - but there is no information on input pin states

If no pins are asserted the `Pn:` element is not output. Which is the same behaviour as legacy Grbl.

**Reply by tcurdt** [reply]
Thanks for the quick help. I was looking for this mainly for debugging purposes.

I assume I can see via `$I` whether my current firmware supports expressions, right? `NEWOPT`?

**Reply by terjeio** [reply]
> I assume I can see via $I whether my current firmware supports expressions, right?

Correct, look for the `EXPR` string.

![image](https://github.com/grblHAL/core/assets/20260062/0864c321-572a-47a4-bd41-79a4c36ac431)
